test/functional/test_framework/util.py

```diff
@@ -47,20 +47,42 @@ def assert_fee_amount(fee, tx_size, feerate_BTC_kvB):
         raise AssertionError("Fee of %s BTC too high! (Should be %s BTC)" % (str(fee), str(target_fee)))
 
 
-def assert_equal(thing1, thing2, *args):
+def assert_equal(thing1, thing2, *args, err_msg=None):
     if thing1 != thing2 or any(thing1 != arg for arg in args):
-        raise AssertionError("not(%s)" % " == ".join(str(arg) for arg in (thing1, thing2) + args))
+        if err_msg is None:
+            err_msg = "not(%s)" % " == ".join(str(arg) for arg in (thing1, thing2) + args)
+        raise AssertionError(err_msg)
 
+def assert_not_equal(thing1, thing2, *args, err_msg=None):
+    if thing1 == thing2 or any(thing1 == arg for arg in args):
+        msg = "%s" % " == ".join(str(arg) for arg in (thing1, thing2) + args)
+        if err_msg is not None:
+            msg = err_msg
+        raise AssertionError(msg)
 
-def assert_greater_than(thing1, thing2):
+
+def assert_greater_than(thing1, thing2, err_msg=None):
     if thing1 <= thing2:
-        raise AssertionError("%s <= %s" % (str(thing1), str(thing2)))
+        msg = "%s <= %s" % (str(thing1), str(thing2))
+        if err_msg is not None:
+            msg = err_msg
+        raise AssertionError(msg)
 
 
-def assert_greater_than_or_equal(thing1, thing2):
+def assert_greater_than_or_equal(thing1, thing2, err_msg=None):
     if thing1 < thing2:
-        raise AssertionError("%s < %s" % (str(thing1), str(thing2)))
-
+        msg = "%s < %s" % (str(thing1), str(thing2))
+        if err_msg is not None:
+            msg = err_msg
+        raise AssertionError(msg)
+
+
+def assert_true(thing, err_msg=None):
+    if thing is not True:
```

Comment:
These functions are generally (e.g. https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertTrue) used to do a _boolean_ comparison, which makes it functionally different than using `assert_equal(expr, True)`

It wouldn't make a functional difference in the places where it's currently used (where a simple `assert` would suffice), but otherwise, this is how I think it should be done:

```suggestion
    if bool(thing) is not True:
```

Reply:
with this suggestion, you will create an error a runtime and no static checker can find it. We can use the only good thing of python to write something like that

```
def assert_true(thing: bool, err_msg=None):
    if thing is not True:
```

In this way, the editor is able to catch the type mismatch! the cast of a type at runtime should be an antipattern in all modern language

Reply:
The only way `bool(thing)` throws an error is if `thing` is of a custom type that explicitly overrides `__bool__()` to return something other than `True` or `False`, which is... really bad code. I don't think we need to handle that, and we can easily observe if it happens from the stacktrace.

Even though I'm a fan of type hints, I don't think they change anything here - afaik we don't use tooling like `mypy` that would enforce them? It also wouldn't do the boolean comparison which was kinda my point: you don't want `assert_true()` to just verify if something is equal to True/False (`assert_equal` can do that), it's helpful if you need to check if _the boolean of_ an object would evaluate to True/False. For example:

```python
result = some_func()  # some_func returns 0 if it failed, an int between 1-5 if successful
assert_equal(result, True) # this would not work, because e.g. `3 == True` would fail
assert_true(result)  # this would work if `assert_true` evaluates bool(result), because `bool(3) is True` would pass

if(result):
    do_something_else()
```

Reply:
But to reiterate, I'm advocating against implementing `assert_true()` at this moment, so I'll leave it here

Reply:
You are right! The thing is worst than the exception because happens something like that

```
➜  ~ python3
Python 3.8.10 (default, Jun 22 2022, 20:18:18) 
[GCC 9.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> bool(12)
True
>>> bool("Alibaba")
True
>>> 

```

This can not hide bugs?

Reply:
I think to make all happy here is to use the assert_equal everywhere it is possible and move on.

This is code that I wrote one year ago, so I do not remember why I insert `assert_true`, but if I was not drunk I think because somewhere there is some comparison with custom error messages that check a boolean condition.

Anyway, I think if you agree we can move all on `assert_equal` and move on.

>you don't want assert_true() to just verify if something is equal to True/False

To sustain your thesis, I'm not able to see any real use case other than toy example of how this can be used

Reply:
Mh! I do not argue this on the PR because it is the wrong place but I disagree with this statement

>How is that worse? You just need to use the right tool for the job.

Of course, bug is a bug, and it is there because there is a mistake in the code, so if a comparison between bool, I want to compare boolean value and not a string that always evaluates to true except for empty.

Is worse because if we were able to catch this bug easily we were still coding in C without struggling with rust compiler, if you do this case you should define some rules like: `bool("False")` -> False or True?

Image to have `assert_true(json['foo'])`  and make an API change that move the type of a filed from string to int

you fall in case of 

```
Python 3.8.10 (default, Jun 22 2022, 20:18:18) 
[GCC 9.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> bool(0)
False
>>> bool("0")
True
>>> bool(0.0)
False
>>> bool("0.0")
True
```

It is not the right tool anyway at least without a good reason to allow this, and I do not see any till now

Reply:
A better example can be

take the from our `assert_true(hash_str is None)` with `assert_true(hash_str)`

```
➜  ~ python3
Python 3.8.10 (default, Jun 22 2022, 20:18:18) 
[GCC 9.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> bool(None)
False
>>> bool("")
False
```

This can be a bug not caught by the CI, and if it is an API change you can break the client that you strongly typed language